lib/fluent/plugin/out_exec.rb

```diff
@@ -18,97 +18,82 @@
-      f
+
+    config_section :inject do
+      config_set_default :time_type, :string
+      config_set_default :localtime, false
     end
-    config_param :localtime, :bool, default: false
-    config_param :timezone, :string, default: nil
 
-    def compat_parameters_default_chunk_key
-      'time'
+    config_section :buffer do
+      config_set_default :delayed_commit_timeout, 300 # 5 min
     end
 
-    def configure(conf)
-      compat_parameters_convert(conf, :buffer, default_chunk_key: 'time')
+    attr_reader :formatter # for tests
 
+    def configure(conf)
+      compat_parameters_convert(conf, :inject, :formatter, :buffer, default_chunk_key: 'time')
       super
-
-      @formatter = case @format
-                   when :tsv
-                     if @keys.empty?
-                       raise Fluent::ConfigError, "keys option is required on exec output for tsv format"
-                     end
-                     Fluent::ExecUtil::TSVFormatter.new(@keys)
-                   when :json
-                     Fluent::ExecUtil::JSONFormatter.new
-                   when :msgpack
-                     Fluent::ExecUtil::MessagePackFormatter.new
-                   end
-
-      if @time_key
-        if @time_format
-          tf = Fluent::TimeFormatter.new(@time_format, @localtime, @timezone)
-          @time_format_proc = tf.method(:format)
-        else
-          @time_format_proc = Proc.new { |time| time.to_s }
-        end
-      end
+      @formatter = formatter_create
     end
 
     def format(tag, time, record)
-      out = ''
-      if @time_key
-        record[@time_key] = @time_format_proc.call(time)
-      end
-      if @tag_key
-        record[@tag_key] = tag
-      end
-      @formatter.call(record, out)
-      out
-    end
-
-    def write(chunk)
-      if chunk.respond_to?(:path)
-        prog = "#{@command} #{chunk.path}"
+      record = inject_values_to_record(tag, time, record)
+      if @formatter.formatter_type == :text_per_line
+        @formatter.format(tag, time, record).chomp + "\n"
       else
-        tmpfile = Tempfile.new("fluent-plugin-exec-")
-        tmpfile.binmode
-        chunk.write_to(tmpfile)
-        tmpfile.close
-        prog = "#{@command} #{tmpfile.path}"
+        @formatter.format(tag, time, record)
       end
+    end
 
-      system(prog)
-      ecode = $?.to_i
-      tmpfile.delete if tmpfile
-
-      if ecode != 0
-        raise "command returns #{ecode}: #{prog}"
-      end
+    def try_write(chunk)
+      tmpfile = nil
+      prog = if chunk.respond_to?(:path)
+               "#{@command} #{chunk.path}"
+             else
+               tmpfile = Tempfile.new("fluent-plugin-exec-")
+               tmpfile.binmode
+               chunk.write_to(tmpfile)
+               tmpfile.close
+               "#{@command} #{tmpfile.path}"
+             end
+      chunk_id = chunk.unique_id
+      callback = ->(status){
+        begin
+          if tmpfile
```

Comment:
No. `ensure` of `#try_write` doesn't take care about child process status.
This callback is called after child process plugin helper find the child process's exit status. So here is the only place to delete tmpfile.